index.js

```diff
@@ -1,60 +1,82 @@
 'use strict'
 
-const pLimit = require('p-limit')
+const assert = require('assert')
 
-async function forEach (iterator, func, n = 16) {
-  const limit = pLimit(n)
-  const promises = new Set()
-  let _err
-  for await (const chunk of iterator) {
-    if (_err) {
-      throw _err
-    }
-    const p = limit(func, chunk)
+async function * mapIterator (iterator, func, n = 16) {
+  const promises = []
 
-    if (limit.pendingCount > 1) {
-      await p
-    } else {
-      promises.add(p)
-      // fork the promise chain
-      p
-        .catch((err) => { _err = err })
-        .finally(() => {
-          promises.delete(p)
-        })
-    }
-  }
+  let next
+  let done = false
 
-  await Promise.all(promises)
-}
+  async function pump () {
+    try {
+      for await (const item of iterator) {
+        if (done) {
+          return
+        }
 
-async function * mapIterator (iterator, func, n = 16) {
-  const limit = pLimit(n)
-  let promises = []
-  let _err
-  for await (const item of iterator) {
-    if (_err) {
-      throw _err
-    }
-    const p = limit(func, item)
+        let p
+        try {
+          p = func(item)
+        } catch (err) {
+          p = Promise.reject(err)
+        }
 
-    promises.push(p)
-    // fork the promise chain
-    p
-      .catch((err) => { _err = err })
+        promises.push(p)
+        p.catch(() => {
+          done = true
+        })
+
+        if (next) {
+          next()
+          next = null
+        }
 
-    if (limit.pendingCount > 1) {
-      for (const promise of promises) {
-        yield await promise
+        if (!done && promises.length >= n) {
+          await new Promise(resolve => {
+            next = resolve
+          })
+          assert(done || promises.length < n)
+        }
+      }
+    } finally {
+      done = true
+      if (next) {
+        next()
+        next = null
       }
-      promises = []
     }
   }
```

Comment:
This promise might reject, could you add a catch?

Reply:
Fixed